File: doc-examples/mongo-example-records-java/src/test/java/example/SaleRepositorySpec.java
package example;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@MicronautTest
class SaleRepositorySpec {

    @Inject ProductRepository productRepository;
    @Inject SaleRepository saleRepository;
    @Inject ManufacturerRepository manufacturerRepository;

    @Test
    void testReadWriteCustomType() {
        Manufacturer apple = manufacturerRepository.save("Apple");
        Product macBook = new Product("MacBook", apple);
        macBook = productRepository.save(macBook);

        Sale sale = saleRepository.save(new Sale(macBook, Quantity.valueOf(1)));

        assertNotNull(
            sale.id()
        );
        assertEquals(1, sale.quantity().getAmount());

        sale = saleRepository.findById(sale.id()).orElse(sale);
        assertNotNull(sale);
        assertNotNull(sale.product());
        assertEquals(1, sale.quantity().getAmount());
        assertTrue(saleRepository.findByQuantity(sale.quantity()).isPresent());
    }
}

File: doc-examples/jdbc-example-kotlin/src/test/kotlin/example/SaleRepositorySpec.kt
package example

import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test

@MicronautTest
class SaleRepositorySpec(
        private val productRepository: ProductRepository,
        private val saleRepository: SaleRepository,
        private val manufacturerRepository: ManufacturerRepository) {

    @Test
    fun testReadWriteCustomType() {
        val apple = manufacturerRepository.save("Apple")
        val macBook = Product(0,"MacBook", apple)
        productRepository.save(macBook)

        var sale = saleRepository.save(Sale(0, macBook, Quantity(1)))

        assertNotNull(
                sale.id
        )
        assertEquals(1, sale.quantity.amount)

        sale = saleRepository.findById(sale.id!!) ?: sale
        assertNotNull(sale)
        assertEquals(1, sale.quantity.amount)
    }
}
